feat: Convenience methods for NestedSet {get_parent, get_children}

barredterra opened the pull request:
Make working with tree DocTypes easier.

Let's take the ERPNext Chart of Accounts as an example:

- 4000 - Income
    - 4100 - Direct Income
        - 4110 - Sales (<- we start here)
        - 4120 - Service
    - 4200 - Indirect Income

```py
doc = frappe.get_doc("Account", "4110 - Sales")

for child in doc.get_children():
    print(child.name)
# <empty>

doc.get_parent().name
# 4100 - Direct Income

for child in doc.get_parent().get_children():
    print(child.name)
# 4110 - Sales
# 4120 - Service

doc.get_parent().get_parent().name
# 4000 - Income

for child in doc.get_parent().get_parent().get_children():
    print(child.name)
# 4100 - Direct Income
# 4200 - Indirect Income

doc.get_parent().get_parent().get_parent()
# None
```

Docs: https://github.com/frappe/frappe_docs/pull/223

barredterra committed:
feat: new properties `NestedSet.{parent, children}`

barredterra committed:
fix: typo

barredterra committed:
refactor: remove unused kwargs

leela commented:
@barredterra, these are very much needed methods for trees. Please add docs so that we can merge the PR.

gavindsouza commented:
Since each invocation of these properties makes database calls, how about `get_parent` or `get_children` instead? 

Also, I don't think we're reserving `parent` and `children` attributes/fieldnames for DocType creations, although we seem to use `parent` quite a lot. 

If we're going ahead with this, I suggest we add a validation at the very least. Maybe via https://github.com/frappe/frappe/blob/713fba3d625ee6905ad3cf335431d2e33bc48d4f/frappe/core/doctype/doctype/doctype.py#L93

barredterra commented:
Hi @gavindsouza, by using `get_cached_doc` the database call will only happen once, when the property is first accessed. The following access will be almost instantaneous.

I guess `get_parent()` and `get_children()` would also be okay, just not as sweet to write and read.

Reserving `parent` and `children` (if we stick to using properties) would surely be a good idea.

gavindsouza commented:
> Hi @gavindsouza, by using `get_cached_doc` the database call will only happen once, when the property is first accessed. The following access will be almost instantaneous.

There are some data consistency concerns with using `get_cached_doc` 😅 .  They get cleared on `set_value`, updated on `get_doc`, not on raw queries or rollbacks. The last bit may raise issues that may be hard to debug. This could work for specific deployments but I think the Framework should be more mindful in these cases.

> I guess `get_parent()` and `get_children()` would also be okay, just not as sweet to write and read.
> 
> Reserving `parent` and `children` (if we stick to using properties) would surely be a good idea.

I suggested the `get_*` methods because accessing these properties trigger some I/O, database or redis. My initial thoughts were to use functools' `cached_property`...but unsure if it respects multi-tenancy. 

A suggestion I had for fetching children would've been to use generators, with get_doc instead of returning a list. It would work better when we have a single node having hundreds of child nodes...specially given a single `get_doc` for a Tree doctype triggers 6-7 queries. But, would be nicer if we got more views on this.

barredterra committed:
feat: method instead of property

barredterra committed:
feat: use generator

barredterra commented:
@gavindsouza refactored as per your suggestion. Any other changes required or can I go ahead with the test cases and docs?

gavindsouza commented:
> @gavindsouza refactored as per your suggestion. Any other changes required or can I go ahead with the test cases and docs?

This is fine i guess. Don't forget to add type hints too 😄

barredterra commented on frappe/utils/nestedset.py:
```diff
@@ -262,6 +262,19 @@ def validate_ledger(self, group_identifier="is_group"):
+			return frappe.get_doc(self.doctype, parent_name)
+
+	def get_children(self):
+		"""Return a list of child Documents."""
```
```suggestion
		"""Return a generator that yields child Documents."""
```

barredterra committed:
docs: fix docstring to match modified method

barredterra committed:
refactor: add type hints

gavindsouza commented:
Hey @barredterra, can you fix the failing tests?

barredterra commented:
Will do, but I'm busy with other things currently. Might take a couple of weeks until I can get back at this PR.

barredterra committed:
refactor: type hints as string for compatibility with python <3.7

https://stackoverflow.com/questions/33533148/how-do-i-type-hint-a-method-with-the-type-of-the-enclosing-class

barredterra committed:
refactor: import Iterator for type annotation

gavindsouza commented on frappe/utils/nestedset.py:
```diff
@@ -10,6 +10,8 @@
 # 3. call update_nsm(doc_obj) in the on_upate method
 
 # ------------------------------------------
+from collections.abc import Iterator
```
```suggestion
from typing import Iterator
```

Do you mean this?

barredterra committed:
refactor: Apply suggestions from code review

Co-authored-by: gavin <gavin18d@gmail.com>

gavindsouza commented:
@barredterra Do we need tests for this?
